modules/core/src-js/smithy4s/Timestamp.scala

```diff
@@ -290,70 +290,86 @@ object Timestamp {
-        || ch4 != '-'
       ) error()
-      pos += 5
-      ch0 * 1000 + ch1 * 100 + ch2 * 10 + ch3 - 53328 // 53328 == '0' * 1111
+      pos += 4
+      year
     }
     val month = {
-      val ch0 = s.charAt(pos)
-      val ch1 = s.charAt(pos + 1)
-      val ch2 = s.charAt(pos + 2)
+      val separator = s.charAt(pos)
+      if (separator != '-' || pos + 3 > len) error()
+      val ch0 = s.charAt(pos + 1)
+      val ch1 = s.charAt(pos + 2)
       val month = ch0 * 10 + ch1 - 528 // 528 == '0' * 11
       if (
-        ch0 < '0' || ch0 > '1' || ch1 < '0' || ch1 > '9' || month < 1 || month > 12 || ch2 != '-'
+        ch0 < '0' || ch0 > '1' || ch1 < '0' || ch1 > '9' || month < 1 || month > 12
       ) error()
       pos += 3
       month
     }
     val day = {
-      val ch0 = s.charAt(pos)
-      val ch1 = s.charAt(pos + 1)
-      val ch2 = s.charAt(pos + 2)
+      val separator = s.charAt(pos)
+      if (separator != '-' || pos + 3 > len) error()
+      val ch0 = s.charAt(pos + 1)
+      val ch1 = s.charAt(pos + 2)
       val day = ch0 * 10 + ch1 - 528 // 528 == '0' * 11
       if (
-        ch0 < '0' || ch0 > '3' || ch1 < '0' || ch1 > '9' || day == 0 ||
-        (day > 28 && day > maxDayForYearMonth(year, month)) || ch2 != 'T'
+        ch0 < '0' || ch0 > '3' || ch1 < '0' || ch1 > '9' || day < 1 || (day > 28 && day > maxDayForYearMonth(
+          year,
+          month
+        ))
       ) error()
       pos += 3
       day
     }
     val hour = {
-      val ch0 = s.charAt(pos)
-      val ch1 = s.charAt(pos + 1)
-      val ch2 = s.charAt(pos + 2)
+      val separator = s.charAt(pos)
+      if (separator != 'T' || pos + 3 > len) error()
+      val ch0 = s.charAt(pos + 1)
+      val ch1 = s.charAt(pos + 2)
       val hour = ch0 * 10 + ch1 - 528 // 528 == '0' * 11
       if (
-        ch0 < '0' || ch0 > '2' || ch1 < '0' || ch1 > '9' || hour > 23 || ch2 != ':'
+        ch0 < '0' || ch0 > '2' || ch1 < '0' || ch1 > '9' || hour < 0 || hour > 23
       ) error()
       pos += 3
       hour
     }
     val minute = {
-      val ch0 = s.charAt(pos)
-      val ch1 = s.charAt(pos + 1)
-      val ch2 = s.charAt(pos + 2)
-      if (ch0 < '0' || ch0 > '5' || ch1 < '0' || ch1 > '9' || ch2 != ':')
+      val separator = s.charAt(pos)
+      if (separator != ':' || pos + 3 > len) error()
+      val ch0 = s.charAt(pos + 1)
+      val ch1 = s.charAt(pos + 2)
+      val minute = ch0 * 10 + ch1 - 528 // 528 == '0' * 11
+      if (
+        ch0 < '0' || ch0 > '5' || ch1 < '0' || ch1 > '9' || minute < 0 || minute > 59
+      )
         error()
       pos += 3
-      ch0 * 10 + ch1 - 528 // 528 == '0' * 11
+      minute
     }
     val second = {
-      val ch0 = s.charAt(pos)
-      val ch1 = s.charAt(pos + 1)
-      if (ch0 < '0' || ch0 > '5' || ch1 < '0' || ch1 > '9') error()
-      pos += 2
-      ch0 * 10 + ch1 - 528 // 528 == '0' * 11
+      val separator = s.charAt(pos)
```

Comment:
I believe most of the changes before this line are unnecessary to the task at hand. Considering this function was written with the philosophy of **absolute maximum performance** by @plokhotnyuk (who's way, WAY more knowledgeable than me when it comes to low-level optimisation on JVM), I want to be respectful of his contributions and will ask that any change that are not strictly necessary to addressing the problem are reverted, unless Andriy himself approves the changes. 

In the context of this function (and the others he's written), his input on the topic trumps any subjective argument any of us may have.

Reply:
Changes looks perfect! 

Sorry for being here too late :)

Reply:
Thanks! Just to check, which version do you think we should proceed with - before or after https://github.com/disneystreaming/smithy4s/pull/1553/commits/a161b9e5fa64ec6956d27795a2d2a9f3f478248e?

Reply:
Minimized (latest) changes are preferable because them save a few CPU cycles in case of errors.

Reply:
Cool, thanks.